Sources/SPTDataLoaderSwift/Request.swift

```diff
@@ -0,0 +1,287 @@
+/*
+ Copyright (c) 2015-2020 Spotify AB.
+
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+ */
+
+import Foundation
+
+typealias ResponseHandler = (Result<SPTDataLoaderResponse, Error>) -> Void
+public typealias ResponseValidator = (SPTDataLoaderResponse) throws -> Void
+
+/// A wrapper for initiating URL requests and handling responses with optional validation.
+///
+/// The `Request` type provides a set of functions that make it easy to chain a series of
+/// response handlers and validators.
+///
+/// The request is executed upon attachment of the first response handler. If a handler is
+/// attached after the response has been received, it will be immediately invoked with the
+/// existing value.
+public final class Request {
+    private let request: SPTDataLoaderRequest
+    private let executionHandler: () -> SPTDataLoaderCancellationToken?
+
+    init(request: SPTDataLoaderRequest, executionHandler: @escaping () -> SPTDataLoaderCancellationToken?) {
+        self.request = request
+        self.executionHandler = executionHandler
+    }
+
+    private enum State {
+        case initialized
+        case executed(token: SPTDataLoaderCancellationToken)
+        case completed(response: SPTDataLoaderResponse)
+        case failed(error: Error)
+        case cancelled
+    }
+
+    private let accessLock = AccessLock()
+    private var state: State = .initialized
+    private var responseHandlers: [ResponseHandler] = []
+    private var responseValidators: [ResponseValidator] = []
+
+    func addResponseValidator(_ responseValidator: @escaping ResponseValidator) {
+        accessLock.sync {
+            switch state {
+            case .initialized, .executed:
+                responseValidators.append(responseValidator)
+            default:
+                break
+            }
+        }
+    }
+
+    func addResponseHandler(_ responseHandler: @escaping ResponseHandler) {
+        var result: Result<SPTDataLoaderResponse, Error>?
+
+        accessLock.sync {
```

Comment:
Does this lock need to be on this entire switch statement? Seems like you only need it on the sections where you modify `responseHandlers`. Always best to keep the critical sections as short as possible.

Reply:
I believe this does need to encompass the entire switch because `state` also needs exclusive access when being read/written.